File: solutions/go/allergies/1/allergies.go
// Package allergies comment
package allergies

var (
	allergens = map[string]uint{
		"eggs":         1,
		"peanuts":      2,
		"shellfish":    4,
		"strawberries": 8,
		"tomatoes":     16,
		"chocolate":    32,
		"pollen":       64,
		"cats":         128,
	}
	labels = []string{
		"eggs", "peanuts", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats",
	}
)

// Allergies comment
func Allergies(allergies uint) (all []string) {
	for _, label := range labels {
		if allergies&1 == 1 {
			all = append(all, label)
		}
		allergies >>= 1
	}
	return
}

// AllergicTo comment
func AllergicTo(allergies uint, allergen string) bool {
	return allergies&allergens[allergen] == allergens[allergen]
}

File: solutions/elixir/allergies/1/lib/allergies.ex
defmodule Allergies do
  import Bitwise

  @allergies %{
    1 => "eggs",
    2 => "peanuts",
    4 => "shellfish",
    8 => "strawberries",
    16 => "tomatoes",
    32 => "chocolate",
    64 => "pollen",
    128 => "cats"
  }

  @items %{
    "eggs" => 1,
    "peanuts" => 2,
    "shellfish" => 4,
    "strawberries" => 8,
    "tomatoes" => 16,
    "chocolate" => 32,
    "pollen" => 64,
    "cats" => 128
  }

  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags),
    do:
      Enum.reduce(@allergies, [], fn {n, label}, acc ->
        if (flags &&& n) == n,
          do: [label | acc],
          else: acc
      end)

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item),
    do: (flags &&& @items[item]) > 0
end
